### PR TITLE
 xpoa-bug-fix-list: fix bugs from v3.8 & v3.9 & opti-fork-ratio

### PR DESCRIPTION
**Pull request by aucusaga**
## Description
 xpoa-bug-fix-list: fix bugs from v3.8 & v3.9 & opti-fork-ratio

## Type of change

Please delete options that are not relevant.
- [ ] Bug fix (non-breaking change which fixes an issue)
Bug1: next proposer A cannot received proposalMSG from preProposer when preBlock add new validate A.
Bug2: preLeader's QC is covered nil when receive proposalMSG.
Bug3: preLeader's viewMSG is invalid when the receiver(next proposer) updates its validates thoroughly.
...

### CODE REVIEW

File: core/consensus/xpoa/xpoa.go [hunk 1]
```diff
@@ -321,171 +324,134 @@ Again:
...
-// TODO: zq needs to be optimized in future because
-func (xpoa *XPoa) getCurrentValidates(curHeight int64) ([]*cons_base.CandidateInfo, error) {
-	preBlockId, err := xpoa.ledger.QueryBlockByHeight(curHeight - 1)
-	if err != nil {
-		xpoa.lg.Error("xpoa.getCurrentValidates", "getBlock", err)
-		return nil, fmt.Errorf("get block by height err:%v", err)
-	}
-	reader, err := xpoa.utxoVM.GetSnapShotWithBlock(preBlockId)
+// getValidatesByBlockId 根据当前输入blockid，用快照的方式在xmodel中寻找<=当前blockid的最新的候选人值，若无则使用xuper.json中指定的初始值
+func (xpoa *XPoa) getValidatesByBlockId(blockId []byte) ([]*cons_base.CandidateInfo, bool, error) {
```
Comment (qizheng09) [inline core/consensus/xpoa/xpoa.go:361]: Does the return param ok is needed?

File: core/consensus/xpoa/xpoa.go [hunk 2]
```diff
@@ -564,8 +535,14 @@ func (xpoa *XPoa) ProcessBeforeMiner(timestamp int64) (map[string]interface{}, b
 	}
 	res["type"] = TYPE
 	if xpoa.enableBFT {
-		if !xpoa.isFirstBlock(xpoa.ledger.GetMeta().GetTrunkHeight() + 1) {
+		height := xpoa.ledger.GetMeta().GetTrunkHeight() + 1
+		if !xpoa.isFirstBlock(height) {
 			if ok, _ := xpoa.bftPaceMaker.IsLastViewConfirmed(); !ok {
+				// 若view number未更新则先暂停
+				if xpoa.bftPaceMaker.CheckViewNumer(height) {
```
Comment (qizheng09) [inline core/consensus/xpoa/xpoa.go:542]: It's will happened in what kind of situation.

File: core/consensus/common/chainedbft/smr/smr.go [hunk 3]
```diff
@@ -87,6 +90,7 @@ func NewSmr(
 		lk:            &sync.Mutex{},
 		QuitCh:        make(chan bool, 1),
 	}
+	effectiveDelay = effectiveHeight
```
Comment (qizheng09) [inline core/consensus/common/chainedbft/smr/smr.go:93]: What's this mean?

File: core/consensus/common/chainedbft/smr/smr.go [hunk 4]
```diff
@@ -254,12 +253,20 @@ func (s *Smr) ProcessProposal(viewNumber int64, proposalID,
...
+		err = s.voteProposal(qc, s.address, netMsg.GetHeader().GetLogid())
+		if err != nil {
+			s.slog.Error("提前签名error", "logid", netMsg.GetHeader().GetLogid(), "error", err)
+		}
+	}
+	s.slog.Info("ProcessProposal: Send proposal msg", "netMsg", netMsg)
 	opts := []p2p_base.MessageOption{
 		p2p_base.WithBcName(s.bcname),
-		p2p_base.WithTargetPeerAddrs(s.getReplicasURL()),
+		p2p_base.WithTargetPeerAddrs(s.getReplicasURL(validatesInfos)),
```
Comment (qizheng09) [inline core/consensus/common/chainedbft/smr/smr.go:269]: Why transmit validates by param? Does this the same mean with `Smr.validates`.

File: core/consensus/common/chainedbft/smr/smr.go [hunk 5]
```diff
@@ -603,9 +633,9 @@ func (s *Smr) UpdateValidateSets(validates []*cons_base.CandidateInfo) error {
 }
 
 // getReplicasURL return validates urls
-func (s *Smr) getReplicasURL() []string {
+func (s *Smr) getReplicasURL(validates []*cons_base.CandidateInfo) []string {
```
Comment (qizheng09) [inline core/consensus/common/chainedbft/smr/smr.go:636]: If  use param to transmit `validates`. This function does not need to be a member function.

File: core/consensus/common/chainedbft/smr/smr.go [hunk 6]
```diff
@@ -51,6 +51,8 @@ var (
 	ErrGetLocalProposalQC = errors.New("get local proposalQC error")
 )
 
+var effectiveDelay = int64(0)
```
Comment (qizheng09) [inline core/consensus/common/chainedbft/smr/smr.go:54]: Why use var for `effectiveDelay `.

**Review by qizheng09 — APPROVED** [review]

**Review by yucaowang — APPROVED** [review]
LGTM

**Pull request merged**